src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.cs

```
@@ -286,66 +285,58 @@ public int GetGeneration(ObjectId? serviceId)
         }
 
         // private methods
-        private async Task MaintainSizeAsync()
+        private async Task MaintainSizeAsync(CancellationToken cancellationToken)
         {
-            var maintenanceCancellationToken = _maintenanceCancellationTokenSource.Token;
-            while (!maintenanceCancellationToken.IsCancellationRequested)
+            try
             {
-                try
-                {
-                    await PrunePoolAsync(maintenanceCancellationToken).ConfigureAwait(false);
-                    await EnsureMinSizeAsync(maintenanceCancellationToken).ConfigureAwait(false);
-                }
-                catch
+                while (!cancellationToken.IsCancellationRequested)
                 {
-                    // ignore exceptions
+                    try
+                    {
+                        await PrunePoolAsync(cancellationToken).ConfigureAwait(false);
+                        await EnsureMinSizeAsync(cancellationToken).ConfigureAwait(false);
+                    }
+                    catch (MongoConnectionException exception)
```

does it make sense to rely on SDAM error handling and defer all exception to that point? Non `MongoConnectionException` will be ignored there anyway

---

I don't think so. SDAM error handling should be applied only in connectivity area, so I don't see why it should be piggybacked for other possible exceptions.

---

>I don't think so. SDAM error handling should be applied only in connectivity area

I don't think that it's fully true statement. We catch all exceptions and it's up to SDAM how many of them we should handle. Yes, currently we handle only MongoConnectionException there but it's decision that we take on the SDAM level. In other words,  if at some moment we will start looking at more exceptions in SDAM, we won't need to change this logic. In other words (x2), we can simply delegate all/any exceptions handling in connection to the SDAM level

---

agreed, I missed that we can catch more exception here (unrelated to connection at all)